s3torchconnector/src/s3torchconnector/s3reader.py

```diff
@@ -53,6 +53,34 @@ def prefetch(self) -> None:
         if self._stream is None:
             self._stream = self._get_stream()
 
+    def readinto(self, buf) -> int:
+        """Read up to len(buf) bytes into a pre-allocated, writable bytes-like object buf.
+        Return the number of bytes read. If no bytes are available, zero is returned.
+
+        Args:
+            buf : writable bytes-like object
+
+        Returns:
+            int : numer of bytes read or zero, if no bytes available
+        """
+        buf_size = len(buf)
+        if self._position_at_end() or buf_size == 0:
+            # If no bytes are available or no place to write data, zero should be returned
+            return 0
+
+        self.prefetch()
```

Comment:
Generally looks good to me. Only one thing that I can't get the grasp on. 
> readinto method, then PyTorch will provide a pre-allocated buffer for writing that data directly, without the need for creating intermediate copies

How large will this be? Can it be smaller than a part_size?

Reply:
Also, might be a bit of a stretch, but can we also measure/assert memory usage? Since the purpose of the fix was to decrease the amount of copy operations, we should have a way to get a sense of that.

Reply:
Yes, it is possible. The amounts of bytes read by PyTorch is not related to the part_size, it depends on the model structure and how it was compressed. Based on the patter I saw during testing, PyTorch will read header block that is less than 100 bytes and then read the block itself that could be bigger or smaller than part_size.

Reply:
Two more things:
- Do we know if the Lightning checkpoints behave the same? Will they leverage the `readinto` method?
- For testing the completeness of reads, I think we should add e2e tests where we pass a client config varying the part_size to `S3Checkpoint` (and `S3LightningCheckpoint` if that's impacted too).

Reply:
> Do we know if the Lightning checkpoints behave the same? Will they leverage the readinto method?

They should behave the same, because they implementation based on using torch.save/torch.load methods.

>     For testing the completeness of reads, I think we should add e2e tests where we pass a client config varying the part_size to S3Checkpoint (and S3LightningCheckpoint if that's impacted too).

The reading of data using the different part_size is covered by unit tests. I'm not sure that we need to also include that check as part of integration tests as it will increase the time for running e2e even more.

> Also, might be a bit of a stretch, but can we also measure/assert memory usage?

From the logical view, we eliminated one copy. I haven't run the profiler tests on memory, as main goal and main measure of success was time needed for downloading data.

> L.E. Otherwise put, what can we add to our tests to ensure we do not introduce regressions over time?

I think we need to work on that under separate tasks as it requires researching what limits to use to prevent fluky results or to relaxed limits.